// ember-idx-forms/addon/inline/inline.js
import Em from 'ember';
var set = Em.set;
var get = Em.get;
var d = Em.debug;
var fmt = Em.fmt;

/**
* This component is the base component of every inline component,
* It renders the value as a link, clicking on the link produce a save | cancel buttons
* 
* Components should use this component to render the default inline mode while render
* its edit mode in this component's yield.
* It does not conduct the edit mode details since the edit mode of every component is too different.
* 
* @class Inline
* @public
**/
export default Em.Component.extend({
  classNameBindings: ['edit:inline-edit:inline-non-edit'],
  tagName: 'a',
  edit: false,
  //componentClicked: (function(e) {
  //  toggle(this, 'edit')
  // }).on('click'),

  linkClicked: (function() {
    if (!get(this, 'edit')) {
        this.send('toggle');
    }
  }).on('click'),

  actions: {
    save: function() {
        d("Saving %@".fmt(get(this, 'elementId')));
        this.send('toggle');
    },
    cancel: function() {
        d("Canceling %@".fmt(get(this, 'elementId')));
        this.send('toggle');
    },
    toggle: function() {
        d("Toggling %@".fmt(get(this, 'elementId')));
        this.toggleProperty('edit');
    }
  }
});
